python/lsst/base/packages.py

```diff
@@ -0,0 +1,258 @@
+        except:
+            pass  # It's not available, so don't care
+
+    packages = {"python": sys.version}
+    # Not iterating with sys.modules.iteritems() because it's not atomic and subject to race conditions
+    moduleNames = sys.modules.keys()
+    for name in moduleNames:
+        module = sys.modules[name]
+        try:
+            ver = getVersionFromPythonModule(module)
+        except:
+            continue  # Can't get a version from it, don't care
+
+        # Remove "foo.bar.version" in favor of "foo.bar"
+        # This prevents duplication when the __init__.py includes "from .version import *"
+        found = False
+        for ending in (".version", "._version"):
+            if name.endswith(ending):
+                name = name[:-len(ending)]
+                if name in packages:
+                    assert ver == packages[name]
+                    found = True
+            elif name in packages:
+                assert ver == packages[name]
+
+        # Use LSST package names instead of python module names
+        # This matches the names we get from the environment (i.e., EUPS) so we can clobber these build-time
+        # versions if the environment reveals that we're not using the packages as-built.
+        if "lsst" in name:
+            name = name.replace("lsst.", "").replace(".", "_")
+
+        packages[name] = ver
+
+    return packages
+
+_eups = None  # Singleton Eups object
+def getEnvironmentPackages():
+    """Provide a dict of products and their versions from the environment
+
+    We use EUPS to determine the version of certain products (those that don't provide
+    a means to determine the version any other way) and to check if uninstalled packages
+    are being used. We only report the product/version for these packages.
+    """
+    try:
+        from eups import Eups, Product
+    except:
+        from lsst.pex.logging import getDefaultLog
+        getDefaultLog().warn("Unable to import eups, so cannot determine package versions from environment")
+        return {}
+
+    # Cache eups object since creating it can take a while
+    global _eups
+    if not _eups:
+        _eups = Eups()
+    products = _eups.findProducts(tags=["setup"])
+
+    # Get versions for things we can't determine via runtime mechanisms
+    # XXX Should we just grab everything we can, rather than just a predetermined set?
+    packages = {prod.name: prod.version for prod in products if prod in ENVIRONMENT}
+
+    # The string 'LOCAL:' (the value of Product.LocalVersionPrefix) in the version name indicates uninstalled
+    # code, so the version could be different than what's being reported by the runtime environment (because
+    # we don't tend to run "scons" every time we update some python file, and even if we did sconsUtils
+    # probably doesn't check to see if the repo is clean).
+    for prod in products:
+        if not prod.version.startswith(Product.Product.LocalVersionPrefix):
```

Comment:
Hmph.  It broke on Jenkins without both of them.  I'll see if I can find a middle road.